pulsar-client-messagecrypto-bc/src/main/java/org/apache/pulsar/client/impl/crypto/MessageCryptoBc.java

```diff
@@ -106,28 +80,64 @@ public class MessageCryptoBc implements MessageCrypto<MessageMetadata, MessageMe
 
 
     private static final SecureRandom secureRandom;
+
+    private static final String providerName;
+    private static BcVersionSpecificCryptoUtility bcVersionSpecificCryptoUtilityDelegate;
+
     static {
-        SecureRandom rand = null;
-        try {
-            rand = SecureRandom.getInstance("NativePRNGNonBlocking");
-        } catch (NoSuchAlgorithmException nsa) {
-            rand = new SecureRandom();
-        }
 
-        secureRandom = rand;
+        providerName = SecurityUtility.getProvider().getName();
+
+        switch (providerName) {
+            case SecurityUtility.BC:
+                SecureRandom rand = null;
+                try {
+                    rand = SecureRandom.getInstance("NativePRNGNonBlocking", providerName);
+                } catch (NoSuchAlgorithmException | NoSuchProviderException nsa) {
+                    rand = new SecureRandom();
+                }
+                secureRandom = rand;
+                break;
+            case SecurityUtility.BC_FIPS:
+                try {
+                    secureRandom = SecureRandom.getInstance("NONCEANDIV", providerName);
+                } catch (NoSuchAlgorithmException | NoSuchProviderException nsa) {
+                    throw new RuntimeException(
+                            "In BC FIPS mode, we expect the specific Random generators to be available!");
+                }
+                break;
```

Comment:
Could you please elaborate on >if "NONCEANDIV" does not work<? 
https://downloads.bouncycastle.org/fips-java/BC-FJA-UserGuide-1.0.2.pdf section 13 explicitly mentions these 2 securerandom generators in BC FIPS.
DEFAULT is the prediction resistant one - so the safest (I think to generate AES keys), and NONCEANDIV is for nonce and IV generation, exactly what we use that random generator when sending messages. But the original implementation was using the on-blocking pseudorandom for both AES key gen and IV generation.

Overall DEFAULT seems to be the stronger one - so depending more on a good entropy source, so DEFAULT may be more prone to blocking if there is not enough entropy.

Maybe what you mean that >if the user cannot give access to a good entropy source, and DEFAULT securerandom is blocking?< - that is definitely worth drilling into and perhaps document that how the default is configured in BC FIPS and what are the assumptions.

Thanks for the deep review!

Reply:
do we have somebody BC expert we can ask to opine? Would be interesting to know if that is possible that `NONCEANDIV` is not available in BC FIPS, and if so is `DEFAULT` a good fallback (i.e. would not be a good chance that that would not be available either).

btw I think your approach absolutely makes sense regarding the choice of securerandoms and I believe that is what I implemented in spirit: for IV generation we use `NONCEANDIV` - as in the doc and the name suggests it's for IV and Nonce generation, and DEFAULT for everything else (AES key generation)

Reply:
> Could we use the NONCEANDIV for generating ASE and sending messages?

We can, but I would advise against it unless an expert says otherwise, as it is not for that based on the documentation. Generally for keys you want a higher-grade random generator (thus `DEFAULT` is prediction resistant). 
A comment referring to this:
https://github.com/indrabasak/bouncycastle-fips-examples/blob/master/src/main/java/com/basaki/bc/fips/random/SecureRandomNumberExample.java#L86-L87

Could you please elaborate on what would we gain from using `NONCEANDIV` for everything?

On the other hand `DEFAULT` is configurable, so I assume if someone wants to use BC FIPS (as it's a very specific industry-related use-case), they should know about how to configure it in case they experience issues.
see: https://www.bouncycastle.org/fips-java/BCFipsIn100.pdf page 13

Reply:
But you are right: if somebody messes up the default secure random config all bets are off, but if somebody attempt that they better know what they are doing.

Reply:
Please let me know if you need help on the PIP and what are the next steps.

thanks!

Reply:
@nodece - I was thinking and this PR is more about fixing a feature what should be there, not sure if it warrants a PIP, and I can just move the new module`bouncy-castle-common` 's content into `pulsar-common` and thus no new module will be required